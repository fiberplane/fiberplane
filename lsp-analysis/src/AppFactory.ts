import { Hono } from "hono";
import type { RouteElement, RouteTree } from "./types";

export class AppFactory {
  private trees: Array<RouteTree>;
  private apps: Map<string, Hono> = new Map();
  private history: Map<string, RouteElement>;

  constructor(trees: Array<RouteTree>) {
    this.trees = trees;
  }

  public resetHistory() {
    this.history = new Map();
  }

  public hasVisited(id: string): boolean {
    return this.history.has(id);
  }

  public getHistoryLength(): number {
    return this.history.size;
  }

  public getVisitedNode(id: string) {
    return this.trees.find((tree) => tree.id === id);
  }

  public getApp(entryId: string): Hono {
    const root = this.trees.find((tree) => tree.id === entryId);
    this.resetHistory();

    if (!root) {
      throw new Error("Could not find root tree");
    }

    return this.createApp(root);
  }

  private createApp(tree: RouteTree) {
    const app = new Hono();

    this.apps.set(tree.id, app);

    app.use((c, next) => {
      this.history.set(tree.id, tree);
      return next();
    });

    // for (const entry of tree.entries) {
    //   switch (entry.type) {
    //     case "ROUTE_ENTRY": {
    //       app.get(entry.path, (c) => {
    //         this.history.set(entry.id, entry);
    //         return c.text("Ok");
    //       });
    //       continue;
    //     }

    //     case "ROUTE_TREE_REFERENCE": {
    //       if (this.apps.has(entry.targetId)) {
    //         const targetApp = this.apps.get(entry.targetId);
    //         app.route(entry.path, targetApp);
    //       } else {
    //         const tree = this.trees.find((tree) => tree.id === entry.targetId);
    //         if (tree) {
    //           const targetApp = this.createApp(tree);
    //           app.route(entry.path, targetApp);
    //         }
    //       }
    //       continue;
    //     }

    //     case "MIDDLEWARE_ENTRY": {
    //       const middlewareFunc = async (c, next) => {
    //         this.history.set(entry.id, entry);
    //         await next();
    //       };
    //       if (entry.path) {
    //         app.use(entry.path, middlewareFunc);
    //       } else {
    //         app.use(middlewareFunc);
    //       }
    //       continue;
    //     }

    //     default: {
    //       const exhaustiveCheck: never = entry;
    //       // Typescript should never happen
    //       throw new Error(`Unsupported entry type: ${exhaustiveCheck}`);
    //     }
    //   }
    // }

    return app;
  }
}
