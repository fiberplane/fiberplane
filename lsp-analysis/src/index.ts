export { Watcher } from "./watcher";
export { setupMonitoring } from "./setup";
