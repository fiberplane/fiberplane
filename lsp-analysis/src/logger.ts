import { createConsola } from "consola";

export const logger = createConsola({
  level: 4,
});
