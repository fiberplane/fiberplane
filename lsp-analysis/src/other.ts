import { Hono } from "hono";
export { Hono as App } from "hono";
// export const App = Hono;
const app = new Hono();
console.log("app", app);
// const app2 = new Hono();
// export type SuperHono = Hono & { super: boolean };
// export const App = Hono;
