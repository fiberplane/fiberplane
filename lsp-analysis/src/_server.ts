/**
 * This file contains the TypeScript language server instance and related functions.
 * It uses the `vscode-jsonrpc` library to create a connection to the language server
 * and handles the initialization, notification registration, and cleanup.
 *
 * Note that the language server is initialized in the user's project directory by spawning a child process,
 * and running `npx typescript-language-server --stdio`.
 * We interface with that process via stdin/stdout.
 */

import { type ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { resolve } from "node:path";
import {
  type MessageConnection,
  StreamMessageReader,
  StreamMessageWriter,
  createMessageConnection,
} from "vscode-jsonrpc/node.js";
import type {
  ConfigurationParams,
  DidChangeConfigurationParams,
  //   // createMessageConnection,
  InitializeParams,
} from "vscode-languageserver-protocol";
import { logger } from "./logger.js";
import { isPublishDiagnosticsParams } from "./types";
import { getFileUri } from "./utils";
// import { connect } from "node:http2";

// let tsServerInstance: {
//   connection: MessageConnection;
//   tsServer: ChildProcessWithoutNullStreams;
// } | null = null;

/**
 * Retrieves the TypeScript language server instance for a given project path.
 * If an instance already exists, it will be reused to avoid unnecessary initialization.
 *
 * @param {string} pathToProject - The path to the project directory.
 * @returns {Promise<{ connection: MessageConnection; tsServer: ChildProcessWithoutNullStreams }>} The TypeScript language server instance.
 */
export async function getTSServer(pathToProject: string) {
  const resolvedPath = resolve(pathToProject);

  // if (tsServerInstance) {
  //   // logger.debug(
  //   //   chalk.dim(
  //   //     `[debug] Reusing existing TS Server instance for project: ${resolvedPath}`,
  //   //   ),
  //   // );
  //   return tsServerInstance;
  // }

  return await initializeTSServer(resolvedPath);
  // return tsServerInstance;
}

async function initializeTSServer(pathToProject: string) {
  logger.debug(`Initializing TS Server for project: ${pathToProject}`);

  const tsServer = spawn(
    "npx",
    [
      "typescript-language-server",
      "--log-level",
      "verbose",
      // "--logVerbosity=verbose",
      "--stdio",
      // "--tsserver-log-file",
      // "out.log",
    ],
    {
      // NOTE - This will add quite a bit of startup time if the user has not yet downloaded typescript-language-server dependency before via npx...
      //        And I haven't tested what that overhead is...
      // NOTE - We *could* run the language server in the project root (via cwd switch), but I think we handle this by initializing the language server
      //        and pointing it to the project root below anyhow.
      // cwd: pathToProject,
      shell: true,
    },
  );

  // Terminate the language server when the Node.js process exits
  process.on("exit", () => {
    logger.debug(
      `[debug] Terminating TS Server instance for project: ${pathToProject}`,
    );
    tsServer.kill();
  });

  // NOTE - Uncomment to debug raw output of ts-language-server
  //
  tsServer.stdout.on("data", (data) => {
    // console.log("<TS-SERVER-DATA>");
    console.log(`--> tsServer stdout: ${data.toString()}`);
    // logger.box(`tsServer stdout: ${data.toString()}`);
    // console.log("</TS-SERVER-DATA>");
  });

  // tsServer.stderr.on("data", (data) => {
  //   logger.error(`tsserver stderr: ${data.toString()}`);
  // });

  tsServer.stdout.on("data", (data) => {
    logger.trace(`tsserver stdout: ${data.toString()}`);
  });

  const connection = createMessageConnection(
    new StreamMessageReader(tsServer.stdout),
    new StreamMessageWriter(tsServer.stdin),
  );

  connection.listen();

  tsServer.on("close", (code) => {
    logger.debug(`typescript-language-server process exited with code ${code}`);
  });

  try {
    const rootUri = getFileUri(pathToProject);
    logger.debug(
      `Initializing typescript language server with rootUri: ${rootUri}`,
    );

    // ClientCapabilities
    const params: InitializeParams = {
      processId: process.pid,
      rootUri: rootUri,
      capabilities: {
        workspace: {
          // codeLens: {
          //   // dynamicRegistration: true,
          //   refreshSupport: true,
          // },
          configuration: true,
        },
      },
      workspaceFolders: [{ uri: rootUri, name: "app" }],
      trace: "verbose",

      initializationOptions: {
        typescript: {
          referencesCodeLens: {
            enabled: true,
          },
          implementationsCodeLens: {
            enabled: true,
          },
        },

        preferences: {
          allowIncompleteCompletions: true,
          includeCompletionsForModuleExports: true,
          includeCompletionsWithInsertText: true,
        },
      },
    };

    const ir = await connection.sendRequest("initialize", params);
    console.log("ir", ir);
    await connection.sendNotification("initialized");

    registerNotificationHandlers(connection);
    await new Promise((resolve) => setTimeout(resolve, 200));
    // enable codelens
    const configParams: DidChangeConfigurationParams = {
      settings: {
        asdasdfasd: "erwe",
        // typescript: {
        referencesCodeLens: {
          enabled: true,
        },
        implementationsCodeLens: {
          enabled: true,
        },
        // },
      },
    };
    // const dcc =
    await connection.sendNotification(
      "workspace/didChangeConfiguration",
      configParams,
    );
    // console.log("configUpdate dcc ", dcc);

    // const configQueryParams: ConfigurationParams = {
    //   items: [
    //     {
    //       section: "typescript.referencesCodeLens",
    //     },
    //     {
    //       section: "typescript.implementationsCodeLens",
    //     },
    //   ],
    // };
    // const r = await connection.sendRequest(
    //   "workspace/configuration",
    //   configQueryParams,
    // );
    // console.log("configUpdate", r);

    return { connection, tsServer };
  } catch (error) {
    logger.error("Error initializing TS Server:", error);
    throw error;
  }
}

/**
 * Registers handlers for various notifications from the TypeScript language server.
 *
 * @param {MessageConnection} connection - The connection to the TypeScript language server.
 */
function registerNotificationHandlers(connection: MessageConnection) {
  const logger = console;
  connection.onNotification("textDocument/didOpen", (params) => {
    logger.debug(
      `[textDocument/didOpen] Received definition request: ${JSON.stringify(
        params,
        null,
        2,
      )}`,
    );
  });
  // Listen for diagnostics
  connection.onNotification(
    "textDocument/publishDiagnostics",
    (params: unknown) => {
      if (!isPublishDiagnosticsParams(params)) {
        logger.debug(
          "[debug] Received unexpected params for `textDocument/publishDiagnostics`",
        );
        return;
      }
      const { uri, diagnostics } = params;
      if (diagnostics.length > 0) {
        logger.info(
          `[textDocument/publishDiagnostics] Diagnostics for ${uri}:`,
        );
        for (const diag of diagnostics) {
          logger.info(
            `- [${diag.severity}] ${diag.message} at ${diag.range.start.line}:${diag.range.start.character}`,
          );
        }
      } else {
        logger.debug(
          `[textDocument/publishDiagnostics] No diagnostics for ${uri}.`,
        );
      }
    },
  );

  // Listen for log messages
  connection.onNotification("window/logMessage", (params) => {
    const { type, message } = params;
    logger.info(`[window/logMessage] Log Message [${type}]: ${message}`);
    logger.debug(params);
  });

  // Listen for show messages
  connection.onNotification("window/showMessage", (params) => {
    const { type, message } = params;
    logger.debug(`[window/showMessage] Show Message [${type}]: ${message}`);
  });
}

// type
