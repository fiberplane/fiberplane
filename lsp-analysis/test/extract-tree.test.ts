import * as path from "node:path";
import { expect, test } from "vitest";
import { setupMonitoring } from "../src";
// import type { RouteTree, SourceReference } from "../src/types";

test.each([
  // {
  //   name: "single file",

  //   location: "./test-case/single",
  // },
  // {
  //   name: "multiple files",
  //   location: "./test-case/multiple",
  // },
  // {
  //   name: "module imports",
  //   location: "./test-case/module-imports",
  // },
  // {
  //   name: "barrel files",
  //   location: "./test-case/barrel-files",
  // },
  // // {
  // //   name: "bindings",
  // //   location: "./test-case/bindings",
  // // },
  // {
  //   name: "split routes",
  //   location: "./test-case/split-routes",
  // },
  // {
  //   name: "empty",
  //   location: "./test-case/empty",
  // },
  // {
  //   name: "goose-quotes",
  //   location: "../../examples/goose-quotes",
  // },
  {
    name: "hono factory",
    location: "./test-case/hono-factory",
  },
  // {
  //   name: "api",
  //   location: "../../api",
  // }
])("run test $name with location $location", async ({ location, name }) => {
  const absolutePath = path.join(__dirname, location);
  const { watcher, findHonoRoutes, teardown } = setupMonitoring(absolutePath);
  try {
    watcher.start();
    // findHonoRoutes();
    const start = performance.now();
    const result = findHonoRoutes();
    console.log(
      `Duration for ${name}: ${(performance.now() - start).toFixed(4)}`,
    );
    expect(result).toMatchSnapshot();
    // console.log(result.results[0].entries);
    // console.log(flattenRouteTree(result.results[0], "/api/geese/:id/generate"));
    // console.log(flattenRouteTree(result.results[0]));
    // flatten
  } finally {
    teardown();
  }
});

// function flattenRouteTree(routeTree: RouteTree, path?: string) {
//   type FileInfo = {
//     imports: Array<string>;
//     // TODO handle multiple things on a single line
//     content: Record<number, string>;
//   };
//   const newFileEntry = () => {
//     return {
//       imports: [],
//       content: {},
//     };
//   };

//   const files: Record<string, FileInfo> = {
//     [routeTree.fileName]: newFileEntry(),
//   };
//   const routes: Array<string> = [];

//   for (const entry of routeTree.entries) {
//     if (
//       (path && entry.path !== path) ||
//       entry.type === "ROUTE_TREE_REFERENCE"
//     ) {
//       console.log("Encountered unsupported route tree reference", entry);
//       continue;
//     }

//     if (entry.type === "MIDDLEWARE_ENTRY") {
//       console.log("Encountered unsupported middleware entry", entry);
//       continue;
//     }

//     routes.push(
//       `${routeTree.name}.${entry.method}("${entry.path}", ${entry.sources
//         .map((source) => source.content)
//         .join(",")})`,
//     );

//     for (const routeSources of entry.sources) {
//       const current = files[routeSources.fileName];

//       const imports = Object.keys(routeSources.modules)
//         .map((module) => {
//           return `import { ${routeSources.modules[module].map((importPath) => {
//             return importPath.import;
//           })} } from "${module}";`;
//         })
//         .join("\n");

//       if (imports) {
//         current.imports.push(imports);
//       }

//       routeSources.references.forEach(parseSource);
//     }
//   }

//   function parseSource(sourceReference: SourceReference) {
//     const { fileName, line, content } = sourceReference;
//     const file: FileInfo = files[fileName] || newFileEntry();
//     file.content[line] = content;
//     files[fileName] = file;
//     file.imports.push(
//       Object.keys(sourceReference.modules)
//         .map((module) => {
//           return `import { ${sourceReference.modules[module].map(
//             (importPath) => {
//               return importPath.import;
//             },
//           )} } from "${module}";`;
//         })
//         .join(", "),
//     );
//     sourceReference.references.forEach(parseSource);
//   }

//   function serializeFile(fileName: string) {
//     const file = files[fileName] || newFileEntry();
//     const lines = Object.keys(file.content)
//       .sort()
//       .map((line) => {
//         return `${file.content[line]}`;
//       });

//     return `// ${fileName}
// ${file.imports.join("\n")}

// ${lines.join("\n\n")}`;
//   }

//   return `${serializeFile(routeTree.fileName)}
// import { Hono } from "hono";

// const ${routeTree.name} = new Hono();

// ${routes.join("\n\n")}

// ${Object.keys(files)
//       .filter((file) => file !== routeTree.fileName)
//       .map((file) => serializeFile(file))
//       .join("\n\n")}
// `;
// }
