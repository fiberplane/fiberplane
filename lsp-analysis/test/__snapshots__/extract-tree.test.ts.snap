// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`run test 'barrel files' with location './test-case/barrel-files' 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@407",
          "method": "get",
          "path": "/user/1",
          "sources": [
            {
              "character": 19,
              "content": "cors()",
              "fileName": "index.ts",
              "id": "index.ts@426",
              "line": 14,
              "modules": {
                "hono": [
                  {
                    "import": "cors",
                    "importPath": "hono/cors",
                    "name": "hono",
                    "version": "4.6.6",
                  },
                ],
              },
              "references": [],
            },
            {
              "character": 27,
              "content": "async (c) => {
  // await getUser();
  const user = await getUser();
  return c.json(user);
}",
              "fileName": "index.ts",
              "id": "index.ts@434",
              "line": 14,
              "modules": {},
              "references": [
                {
                  "character": 9,
                  "content": "export { getUser } from "./getUser"",
                  "fileName": "db/index.ts",
                  "id": "db/index.ts@9",
                  "line": 0,
                  "modules": {},
                  "references": [
                    {
                      "character": 23,
                      "content": "const getUser = measure("getUser", async () => {
  await sleep(100);
  const value: User = {
    name: DEFAULT_USER_NAME,
    email: DEFAULT_EMAIL,
  };
  return value;
})",
                      "fileName": "db/getUser.ts",
                      "id": "db/getUser.ts@222",
                      "line": 7,
                      "modules": {
                        "@fiberplane/hono-otel": [
                          {
                            "import": "measure",
                            "importPath": "@fiberplane/hono-otel",
                            "name": "@fiberplane/hono-otel",
                            "version": "0.3.1",
                          },
                        ],
                      },
                      "references": [
                        {
                          "character": 0,
                          "content": "export async function sleep(duration: number) {
  return new Promise((resolve) => setTimeout(resolve, duration));
}",
                          "fileName": "utils.ts",
                          "id": "utils.ts@0",
                          "line": 0,
                          "modules": {},
                          "references": [],
                        },
                        {
                          "character": 0,
                          "content": "export type User = {
  name: string;
  email: Email;
};",
                          "fileName": "types.ts",
                          "id": "types.ts@29",
                          "line": 2,
                          "modules": {},
                          "references": [
                            {
                              "character": 0,
                              "content": "export type Email = string;",
                              "fileName": "types.ts",
                              "id": "types.ts@0",
                              "line": 0,
                              "modules": {},
                              "references": [],
                            },
                          ],
                        },
                        {
                          "character": 6,
                          "content": "const DEFAULT_USER_NAME = "John Doe"",
                          "fileName": "db/getUser.ts",
                          "id": "db/getUser.ts@128",
                          "line": 4,
                          "modules": {},
                          "references": [],
                        },
                        {
                          "character": 6,
                          "content": "const DEFAULT_EMAIL = "john@doe.com"",
                          "fileName": "db/getUser.ts",
                          "id": "db/getUser.ts@166",
                          "line": 5,
                          "modules": {},
                          "references": [],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "index.ts@531",
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "index.ts",
              "id": "index.ts@544",
              "line": 20,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@102",
      "name": "app",
    },
  ],
}
`;

exports[`run test 'module imports' with location './test-case/module-imports' 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@133",
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 2,
              "content": "() => cors()",
              "fileName": "index.ts",
              "id": "index.ts@207",
              "line": 9,
              "modules": {
                "hono": [
                  {
                    "import": "cors",
                    "importPath": "hono/cors",
                    "name": "hono",
                    "version": "4.6.6",
                  },
                ],
              },
              "references": [],
            },
            {
              "character": 2,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "index.ts",
              "id": "index.ts@223",
              "line": 10,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "index.ts@456",
          "method": "get",
          "path": "/slow",
          "sources": [
            {
              "character": 17,
              "content": "cors()",
              "fileName": "index.ts",
              "id": "index.ts@473",
              "line": 19,
              "modules": {
                "hono": [
                  {
                    "import": "cors",
                    "importPath": "hono/cors",
                    "name": "hono",
                    "version": "4.6.6",
                  },
                ],
              },
              "references": [],
            },
            {
              "character": 25,
              "content": "async (c) => {
  await sleep(1000);
  return c.text("Hello, Hono (slow)!");
}",
              "fileName": "index.ts",
              "id": "index.ts@481",
              "line": 19,
              "modules": {},
              "references": [
                {
                  "character": 0,
                  "content": "export async function sleep(duration: number) {
  return new Promise((resolve) => setTimeout(resolve, duration));
}",
                  "fileName": "index.ts",
                  "id": "index.ts@290",
                  "line": 14,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "index.ts@848",
          "method": "get",
          "path": "user/1",
          "sources": [
            {
              "character": 18,
              "content": "cors()",
              "fileName": "index.ts",
              "id": "index.ts@866",
              "line": 32,
              "modules": {
                "hono": [
                  {
                    "import": "cors",
                    "importPath": "hono/cors",
                    "name": "hono",
                    "version": "4.6.6",
                  },
                ],
              },
              "references": [],
            },
            {
              "character": 26,
              "content": "async (c) => {
  // await getUser();
  const user = await getUser();
  return c.json(user);
}",
              "fileName": "index.ts",
              "id": "index.ts@874",
              "line": 32,
              "modules": {},
              "references": [
                {
                  "character": 23,
                  "content": "const getUser = measure("getUser", async () => {
  await new Promise((resolve) => setTimeout(resolve, 100));
  const value: User = {
    name: "John Doe",
    email: "john@doe.com",
  };
  return value;
})",
                  "fileName": "db.ts",
                  "id": "db.ts@137",
                  "line": 4,
                  "modules": {
                    "@fiberplane/hono-otel": [
                      {
                        "import": "measure",
                        "importPath": "@fiberplane/hono-otel",
                        "name": "@fiberplane/hono-otel",
                        "version": "0.3.1",
                      },
                    ],
                  },
                  "references": [
                    {
                      "character": 0,
                      "content": "export type User = {
  name: string;
  email: Email;
};",
                      "fileName": "types.ts",
                      "id": "types.ts@29",
                      "line": 2,
                      "modules": {},
                      "references": [
                        {
                          "character": 0,
                          "content": "export type Email = string;",
                          "fileName": "types.ts",
                          "id": "types.ts@0",
                          "line": 0,
                          "modules": {},
                          "references": [],
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@114",
      "name": "app",
    },
  ],
}
`;

exports[`run test 'multiple files' with location './test-case/multiple' 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@55",
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "index.ts",
              "id": "index.ts@68",
              "line": 4,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@36",
      "name": "app",
    },
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "other.ts@53",
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => {
  return c.json({});
}",
              "fileName": "other.ts",
              "id": "other.ts@66",
              "line": 2,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "other.ts@100",
          "method": "post",
          "path": "/",
          "sources": [
            {
              "character": 14,
              "content": "(c) => c.json({ hello: "world" })",
              "fileName": "other.ts",
              "id": "other.ts@114",
              "line": 5,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "other.ts",
      "id": "other.ts@35",
      "name": "app",
    },
  ],
}
`;

exports[`run test 'single file' with location './test-case/single' 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@55",
          "method": "get",
          "path": "/",
          "sources": [
            {
              "character": 13,
              "content": "(c) => c.text("Hello, Hono!")",
              "fileName": "index.ts",
              "id": "index.ts@68",
              "line": 4,
              "modules": {},
              "references": [],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@36",
      "name": "app",
    },
  ],
}
`;

exports[`run test 'split routes' with location './test-case/split-routes' 1`] = `
{
  "errorCount": 0,
  "results": [
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "users.ts@179",
          "method": "get",
          "path": "/api/v1/users",
          "sources": [
            {
              "character": 27,
              "content": "(c) => {
  return c.json(USERS);
}",
              "fileName": "users.ts",
              "id": "users.ts@206",
              "line": 17,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const USERS = [
  {
    id: 1,
    name: "Alice",
    age: 25,
  },
  {
    id: 2,
    name: "Bob",
    age: 30,
  },
]",
                  "fileName": "users.ts",
                  "id": "users.ts@63",
                  "line": 4,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "users.ts@244",
          "method": "get",
          "path": "/api/v1/users/:id",
          "sources": [
            {
              "character": 31,
              "content": "(c) => {
  const id = Number.parseInt(c.req.param("id"));
  const user = USERS.find((u) => u.id === id);
  return c.json(user);
}",
              "fileName": "users.ts",
              "id": "users.ts@275",
              "line": 21,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const USERS = [
  {
    id: 1,
    name: "Alice",
    age: 25,
  },
  {
    id: 2,
    name: "Bob",
    age: 30,
  },
]",
                  "fileName": "users.ts",
                  "id": "users.ts@63",
                  "line": 4,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "users.ts",
      "id": "users.ts@36",
      "name": "users",
    },
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@320",
          "method": "get",
          "path": "/api/v1/projects",
          "sources": [
            {
              "character": 33,
              "content": "(c) => {
  return c.json(PROJECTS);
}",
              "fileName": "index.ts",
              "id": "index.ts@353",
              "line": 19,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const PROJECTS = [
  {
    id: 1,
    name: "Project A",
    description: "Project A description",
  },
  {
    id: 2,
    name: "Project B",
    description: "Project B description",
  },
]",
                  "fileName": "index.ts",
                  "id": "index.ts@133",
                  "line": 6,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "index.ts@394",
          "method": "get",
          "path": "/api/v1/projects/:id",
          "sources": [
            {
              "character": 37,
              "content": "(c) => {
  const id = Number.parseInt(c.req.param("id"));
  const project = PROJECTS.find((p) => p.id === id);
  return c.json(project);
}",
              "fileName": "index.ts",
              "id": "index.ts@431",
              "line": 23,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const PROJECTS = [
  {
    id: 1,
    name: "Project A",
    description: "Project A description",
  },
  {
    id: 2,
    name: "Project B",
    description: "Project B description",
  },
]",
                  "fileName": "index.ts",
                  "id": "index.ts@133",
                  "line": 6,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@103",
      "name": "projects",
    },
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "index.ts@619",
          "path": "/",
          "sources": [
            {
              "character": 8,
              "content": "cors()",
              "fileName": "index.ts",
              "id": "index.ts@627",
              "line": 32,
              "modules": {
                "hono": [
                  {
                    "import": "cors",
                    "importPath": "hono/cors",
                    "name": "hono",
                    "version": "4.6.6",
                  },
                ],
              },
              "references": [],
            },
          ],
          "type": "MIDDLEWARE_ENTRY",
        },
      ],
      "fileName": "index.ts",
      "id": "index.ts@601",
      "name": "app",
    },
    {
      "baseUrl": "",
      "entries": [
        {
          "id": "projects.ts@253",
          "method": "get",
          "path": "/api/v1/projects",
          "sources": [
            {
              "character": 33,
              "content": "(c) => {
  return c.json(PROJECTS);
}",
              "fileName": "projects.ts",
              "id": "projects.ts@286",
              "line": 17,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const PROJECTS = [
  {
    id: 1,
    name: "Project A",
    description: "Project A description",
  },
  {
    id: 2,
    name: "Project B",
    description: "Project B description",
  },
]",
                  "fileName": "projects.ts",
                  "id": "projects.ts@66",
                  "line": 4,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
        {
          "id": "projects.ts@327",
          "method": "get",
          "path": "/api/v1/projects/:id",
          "sources": [
            {
              "character": 37,
              "content": "(c) => {
  const id = Number.parseInt(c.req.param("id"));
  const project = PROJECTS.find((p) => p.id === id);
  return c.json(project);
}",
              "fileName": "projects.ts",
              "id": "projects.ts@364",
              "line": 21,
              "modules": {},
              "references": [
                {
                  "character": 6,
                  "content": "const PROJECTS = [
  {
    id: 1,
    name: "Project A",
    description: "Project A description",
  },
  {
    id: 2,
    name: "Project B",
    description: "Project B description",
  },
]",
                  "fileName": "projects.ts",
                  "id": "projects.ts@66",
                  "line": 4,
                  "modules": {},
                  "references": [],
                },
              ],
            },
          ],
          "type": "ROUTE_ENTRY",
        },
      ],
      "fileName": "projects.ts",
      "id": "projects.ts@36",
      "name": "projects",
    },
  ],
}
`;
